src/Http/Request/RequestExecutor.ts

```diff
@@ -7,79 +8,152 @@ import {IRavenResponse} from "../../Database/RavenCommandResponse";
 import {StringUtil} from "../../Utility/StringUtil";
 import {DateUtil} from "../../Utility/DateUtil";
 import {IResponse, IErrorResponse} from "../Response/IResponse";
 import {StatusCodes, StatusCode} from "../Response/StatusCode";
 import {Observable} from "../../Utility/Observable";
 import {NodeSelector} from "./NodeSelector";
-import {RavenException, InvalidOperationException, BadRequestException, AuthorizationException, TopologyNodeDownException, AllTopologyNodesDownException} from "../../Database/DatabaseExceptions";
+import {NodeStatus} from "../NodeStatus";
+import {RavenException, InvalidOperationException, BadRequestException, AuthorizationException, TopologyNodeDownException, AllTopologyNodesDownException, DatabaseLoadFailureException, UnsuccessfulRequestException} from "../../Database/DatabaseExceptions";
+
+export interface ITopologyUpdateEvent {
+  topologyJson: object;
+  serverNodeUrl: string;
+  requestedDatabase?: string;
+  forceUpdate?: boolean;
+  wasUpdated?: boolean;
+}
+
+export interface IRequestExecutorOptions {
+  withoutTopology?: boolean;
+  topologyEtag?: number;
+  singleNodeTopology?: Topology;
+  firstTopologyUpdateUrls?: string[];
+}
 
 export class RequestExecutor extends Observable {
   public static readonly REQUEST_FAILED      = 'request:failed';
   public static readonly TOPOLOGY_UPDATED    = 'topology:updated';
   public static readonly NODE_STATUS_UPDATED = 'node:status:updated';
 
-  protected conventions?: DocumentConventions;
   protected headers: IHeaders;
-  private _lock: Lock;
-  private _authenticator: ApiKeyAuthenticator;
-  private _apiKey?: string = null;
-  private _unauthorizedHandlerInitialized: boolean = false;
-  private _firstTopologyUpdateCompleted: boolean = false;
-  private _disableTopologyUpdates: boolean;
+  private readonly _maxFirstTopologyUpdatesTries: number = 5;
+  private _fistTopologyUpdatesTries: number = 0;
+  private _awaitFirstTopologyLock: Lock;
+  private _updateTopologyLock: Lock;
+  private _updateFailedNodeTimerLock: Lock;
+  private _firstTopologyUpdate: BluebirdPromise<void> = null;
+  private _withoutTopology: boolean;
   private _nodeSelector: NodeSelector;
-  private _initialUrls: string[];
+  private _lastKnownUrls: string[];
   private _initialDatabase: string;
+  private _topologyEtag: number;
+  private _faildedNodesStatuses: Map<ServerNode, NodeStatus>;
 
   public get initialDatabase(): string {
     return this._initialDatabase;
   }
 
-  constructor(urls: string[], database: string, apiKey?: string, conventions?: DocumentConventions) {
+  constructor(database: string, options: IRequestExecutorOptions = {}) {
     super();
 
+    const urls = options.firstTopologyUpdateUrls || [];
+
     this.headers = {
       "Accept": "application/json",
-      "Has-Api-key": TypeUtil.isNone(apiKey) ? 'false' : 'true',
-      "Raven-Client-Version": "4.0.0.0",
+      "Has-Api-key": 'false',
+      "Raven-Client-Version": "4.0.1.2",
```

Comment:
Let's keep it in sync with what's in *package.json* with this. Python client has different versioning pattern.